File: 2020/day2PasswordPhilosophy/day2.mjs
// --- Day 2: Password Philosophy ---

import { readFileSync } from "node:fs";

const input = "day2Input.txt";

const passwordsDocument = readFileSync(input, { encoding: "utf-8" })
  .split("\n")
  .map((passwordInfo) =>
    passwordInfo.replace(/-/, " ").replace(/:/, "").split(" ")
  );

let validPasswords = 0;

passwordsDocument.forEach(([lowest, highest, letter, password]) => {
  const reEx = new RegExp(letter, "g");
  const matches = password.match(reEx);
  const appearTimes = matches ? matches.length : 0;
  if (appearTimes >= parseInt(lowest) && appearTimes <= parseInt(highest)) {
    validPasswords += 1;
  }
});

console.log(`There are ${validPasswords} valid passwords`); // Answer Part 1

// Part 2

validPasswords = 0;

passwordsDocument.forEach(
  ([validPosition1Str, validPosition2Str, letter, password]) => {
    const validPosition1 = parseInt(validPosition1Str) - 1;
    const validPosition2 = parseInt(validPosition2Str) - 1;

    if (
      (password[validPosition1] === letter) !==
      (password[validPosition2] === letter)
    ) {
      validPasswords += 1;
    }
  }
);

console.log(
  `According to the new interpretation of the policies are ${validPasswords} valid passwords`
); // Answer Part 2

File: 2020/day2PasswordPhilosophy/day2.py
# --- Day 2: Password Philosophy ---

import re

input = "day2Input.txt"

with open(input, "r", encoding="utf-8") as file:
    passwords_document = [
        re.sub(r"[-:]", " ", password_info).split()
        for password_info in file.read().split("\n")
    ]

valid_passwords = 0

for lowest, highest, letter, password in passwords_document:
    appear_times = str.count(password, letter)
    if int(lowest) <= appear_times <= int(highest):
        valid_passwords += 1

print(f"There are {valid_passwords} valid passwords")  # Answer Part 1

# Part 2

valid_passwords = 0

for valid_position_1, valid_position_2, letter, password in passwords_document:
    valid_position_1, valid_position_2 = (
        int(valid_position_1) - 1,
        int(valid_position_2) - 1,
    )

    if (password[valid_position_1] == letter) != (password[valid_position_2] == letter):
        valid_passwords += 1

print(
    f"According to the new interpretation of the policies are {valid_passwords} valid passwords"
)  # Answer Part 2
